cmd/ltkeycloak/from_mattermost.go

```diff
@@ -0,0 +1,286 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"time"
+
+	gocloak "github.com/Nerzal/gocloak/v13"
+	"github.com/mattermost/mattermost-load-test-ng/deployment"
+	"github.com/mattermost/mattermost-load-test-ng/deployment/terraform"
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/mattermost/mattermost/server/public/shared/mlog"
+	"github.com/spf13/cobra"
+)
+
+func userTxtWriter(usersTxtChan chan string, doneChan chan struct{}, userPassword string) {
+	usersTxtFile, err := os.OpenFile("users.txt", os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		mlog.Error("failed to open users.txt file")
+	}
+	defer usersTxtFile.Close()
+
+	for {
+		select {
+		case email := <-usersTxtChan:
+			_, err := usersTxtFile.Write([]byte(model.UserAuthServiceSaml + ":" + email + " " + userPassword + "\n"))
+			if err != nil {
+				mlog.Error("failed to write to users.txt", mlog.String("err", err.Error()))
+			}
+
+		case <-doneChan:
+			return
+		}
+	}
+}
+
+type workerConfig struct {
+	workerNumber     int
+	usersChan        chan *model.User
+	wg               *sync.WaitGroup
+	mmClient         *model.Client4
+	keycloakClient   *gocloak.GoCloak
+	keycloakToken    *gocloak.JWT
+	keycloakRealm    string
+	userPassword     string
+	deploymentConfig *deployment.Config
+	doneChan         chan struct{}
+}
+
+func migrateMattermostUsersToKeycloak(worker *workerConfig) {
+	refreshTokenTicker := time.NewTicker(30 * time.Second)
+	defer refreshTokenTicker.Stop()
+
+	for {
+		select {
+		case user := <-worker.usersChan:
+			ctx, cancel := context.WithTimeout(context.Background(), operationTimeout)
+
+			kcUserID, err := worker.keycloakClient.CreateUser(ctx, worker.keycloakToken.AccessToken, worker.keycloakRealm, gocloak.User{
+				Username:      &user.Username,
+				Email:         &user.Email,
+				Enabled:       gocloak.BoolP(true),
+				EmailVerified: &user.EmailVerified,
+				Credentials: &[]gocloak.CredentialRepresentation{
+					{
+						Temporary: gocloak.BoolP(false),
+						Type:      gocloak.StringP("password"),
+						Value:     gocloak.StringP(worker.userPassword),
+					},
+				},
+			})
+			if err != nil {
+				cancel()
```

Comment:
Do we need to call cancel here if `worker.keycloakClient.CreateUser` has already returned?

Reply:
We need to call `cancel` when the operation finishes (same as `defer`), that means correctly or incorrectly, in order to free resources. If we don't mind that I can just keep using `defer`.

Reply:
Ah, sorry, you're right, a function could return while its goroutines may still be around listening for the ctxt :facepalm: 
I was gonna suggest to call `defer cancel()` right after its definition so that we can skip all the individual calls to `cancel`, but we need to `continue` in some cases. Would it be worth it to wrap all this logic in a function? I'll leave it to you.

Reply:
Something like https://github.com/mattermost/mattermost-load-test-ng/pull/758/commits/19222db309f5ff41ce6eac099204005d95b8fa74 ?